doc/dvswitch.1
.\" dvswitch.1 written by Ben Hutchings <ben@decadent.org.uk>
.TH DVSWITCH 1 "18 February 2009"
.SH NAME
dvswitch \- mixes and distributes DV streams
.SH SYNOPSIS
.HP
.B dvswitch
.RI [ OPTIONS ]
.SH DESCRIPTION
.LP
DVswitch is a basic video mixer designed for use at conferences that
require live mixing for streams and/or recording.  It presents a GUI
for interactive control and runs a network server for connection of
sources and sinks.
.LP
The GUI is documented in the \fBREADME\fR file.
.SH OPTIONS
\fB\-h\fR, \fB\-\-host=\fIHOST\fR
.TP
\fB\-p\fR, \fB\-\-port=\fIPORT\fR
.RS
Specify the network address on which to listen and accept source
and sink connections.  The host address may be specified by name
or as an IPv4 or IPv6 literal.
.RE
.TP
\fB\-o\fR, \fB\-\-osc=\fIPORT\fR
.RS
This option activates the remote-control mode.
If set, dvswitch will listen for OSC messages on the given port.
OSC messages understood by dvcontrol are documented in the
\fBREADME\fR file.
.RE
.SH AUTHOR
Ben Hutchings <ben@decadent.org.uk>.
.SH SEE ALSO
gtk-options(7), /usr/share/doc/dvswitch/README


ChangeLog
dvswitch (unreleased)
  [ Robin Gareus ]
  * OSC remote-control (see doc/README)
  * add contrib examples and tools for OSC

dvswitch (0.9) unstable; urgency=low

  [ Wouter Verhelst ]
  * Don't try to build or install dvsource-alsa on non-Linux.
  * Revisit the interface to make it slihgtly more intuitive again.

  [ Robin Gareus ]
  * fast-fades (allow fade effect 40ms..15sec)
  * ffmpeg API update (for ffmpeg-0.10)
  * dynamic A/B overlay
  * confirm application quit dialog
  * blinking status bar
  * updated GUI widget alignment, and labels
  * various updates to Debian packaging
  * various spelling and formating fixes (doc, man-pages)

 -- Wouter Verhelst <wouter@debian.org>  Wed, 28 Mar 2012 00:39:04 +0200

dvswitch (0.9~rc1) experimental; urgency=low

  [ Ben Hutchings ]
  * Added dvsource-alsa as an audio-only source
  * Create one encoding thread per processor, up to a maximum of 8
  * Add status overlay to make it more clear whether recording is enabled
  * Set display to black before a source is connected
  * Add static Makefile at top level
  * Override source frame aspect ratio in case of a mismatch, rather then
    warning about it - Alioth feature #311953
  * Fix APTs and validity flags in header block - Alioth bug #311836
  * List all needed libraries to the linker.  We were accidentally
    depending on the behaviour of GNU ld, which treats 'needed' as
    transitive by default - Debian bug #554310
  * Remove support for 44.1 kHz audio
  * Add update-po target for updating translation files
  * Fix build failure with ffmpeg 0.6
  * Fix crash at startup with ffmpeg 0.6 due to a new parameter check
  * Enable selection of Firewire DV devices by GUID, thanks to
    Manuel Virgilio
  * Implement dvsource-firewire-rtsp, to serve DV over RTSP (experimental)

  [ Carl Karsten ]
  * Add peak level indicators to VU-meter

  [ Wouter Verhelst ]
  * Add support for tally lights in dvsource-firewire/dvsource-v4l2-dv
  * Internationalize and add Dutch translation
  * Disable effects buttons until sufficient sources have connected.
    Pressing them before sources are connected would result in a crash.
  * Implement fading on an 8-bit scale
  * Disable record button if we can't record - Alioth bug #312329
  * Add tally light support to dvsource-firewire-rtsp
  * Install dvsource-firewire-rtsp into .../bin
  * Take over day-to-day maintenance from Ben

  [ Robin Gareus ]
  * Add dvsource-jack, similar to dvsource-alsa but for JACK audio sources

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 01 Mar 2009 03:16:30 +0000

dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers
    - Changed selection of visual class and depth for images to match the
      window
    - Added error messages and later checks for failure of image creation,
      where previously dvswitch could crash

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 21 Feb 2008 22:17:32 +0000


src/dvswitch.cpp
// Copyright 2007-2008 Ben Hutchings.
// See the file "COPYING" for licence details.

// Top level of DVswitch

#include <cstdlib>
#include <cstring>
#include <exception>
#include <iostream>
#include <ostream>
#include <string>

#include <getopt.h>

#include <locale.h>
#include <libintl.h>

#include <gtkmm/main.h>
#include <sigc++/functors/slot.h>

#include "avcodec_wrap.h"

#include "config.h"
#include "connector.hpp"
#include "mixer.hpp"
#include "mixer_window.hpp"
#include "server.hpp"
#include "osc_ctrl.hpp"

namespace
{
    struct option options[] = {
	{"host",             1, NULL, 'h'},
	{"port",             1, NULL, 'p'},
	{"osc",              1, NULL, 'o'},
	{"help",             0, NULL, 'H'},
	{NULL,               0, NULL, 0}
    };

    std::string mixer_host;
    std::string mixer_port;

    extern "C"
    {
	void handle_config(const char * name, const char * value)
	{
	    if (std::strcmp(name, "MIXER_HOST") == 0)
		mixer_host = value;
	    else if (strcmp(name, "MIXER_PORT") == 0)
		mixer_port = value;
	}
    }

    void usage(const char * progname)
    {
	std::cerr << "\
Usage: " << progname << " [gtk-options] \\\n\
           [{-h|--host} LISTEN-HOST] [{-p|--port} LISTEN-PORT] [{-o|--osc} OSC-PORT]\n";
    }
}

int main(int argc, char **argv)
{
    setlocale(LC_ALL, "");
    bindtextdomain("dvswitch", SHAREDIR "/locale");
    textdomain("dvswitch");
    try
    {
	dvswitch_read_config(handle_config);

	// Initialise Gtk
	Gtk::Main kit(argc, argv);

	// Complete option parsing with Gtk's options out of the way.

	int osc_port = 0;
	int opt;
	while ((opt = getopt_long(argc, argv, "h:p:o:", options, NULL)) != -1)
	{
	    switch (opt)
	    {
	    case 'h':
		mixer_host = optarg;
		break;
	    case 'p':
		mixer_port = optarg;
		break;
	    case 'o': /* --osc */
		osc_port = atoi(optarg);
		break;
	    case 'H': /* --help */
		usage(argv[0]);
		return 0;
	    default:
		usage(argv[0]);
		return 2;
	    }
	}

	if (mixer_host.empty() || mixer_port.empty())
	{
	    std::cerr << argv[0] << ": mixer hostname and port not defined\n";
	    return 2;
	}

	// The mixer must be created before the window, since we pass
	// a reference to the mixer into the window's constructor to
	// allow it to adjust the mixer's controls.
	// However, the mixer must also be destroyed before the
	// window, since as long as it exists it may call into the
	// window as a monitor.
	// This should probably be fixed by a smarter design, but for
	// now we arrange this by attaching the window to an auto_ptr.
	std::auto_ptr<mixer_window> the_window;
	mixer the_mixer;
	server the_server(mixer_host, mixer_port, the_mixer);
	connector the_connector(the_mixer);
	the_window.reset(new mixer_window(the_mixer, the_connector));
	the_mixer.set_monitor(the_window.get());
	the_window->show();
	the_window->signal_hide().connect(sigc::ptr_fun(&Gtk::Main::quit));
	if (osc_port > 0 )
	{
	    OSC *os = new OSC();
	    if (!os->initialize_osc(osc_port))
	    {
		os->setup_thread(Glib::MainContext::get_default());
		the_window->init_osc_connection(os);
	    }
	}
	Gtk::Main::run();
	return EXIT_SUCCESS;
    }
    catch (std::exception & e)
    {
	std::cerr << "ERROR: " << e.what() << "\n";
	return EXIT_FAILURE;
    }
    catch (Glib::Exception & e)
    {
       std::cerr << "ERROR: " << e.what() << "\n";
       return EXIT_FAILURE;
    }
}
